### PR TITLE
fix(migrations): make the full_path_hash migration re-runnable

### PR DESCRIPTION
**Pull request by sdornan**
**Description**
<sup>Explain the changes or enhancements you are proposing with this pull request.</sup>

Upgrading to the 5.3.0 alpha can fail on every start with:

```
sqlalchemy.exc.OperationalError: (mariadb.OperationalError) Duplicate column name 'full_path_hash'
[SQL: ALTER TABLE roms ADD COLUMN full_path_hash VARCHAR(64)]
```

That error is a symptom, not the cause. MariaDB and MySQL auto-commit each DDL statement, so a run of `0126_unique_rom_full_path` that dies after its `ADD COLUMN` (a container restart, a timeout on the whole-table backfill, or a failure on the `NOT NULL` alter or either index) keeps the column while `alembic_version` stays at `0125`. Every start after that replays `0126` from the top and dies on the first statement, which hides whatever actually stopped the upgrade. The init script logs the failure and starts the app anyway, so the instance keeps running on a half-migrated schema.

`0127` was already written against this hazard ("MySQL/MariaDB auto-commit each DDL statement, so a crash mid-migration leaves objects behind without advancing the alembic version"); `0126` and `0128` were not.

- `0126` now reflects the `roms` schema first and guards each step: add the column only if it is absent, backfill only the rows still missing a digest, apply `NOT NULL` only while the column is nullable, and rebuild `idx_roms_platform_id_fs_name` only while it is still unique. A replay resumes where the previous run stopped, so an interrupted upgrade recovers on the next start and, if the original failure was a real one, it surfaces its own error instead of the duplicate-column one.
- `0128` adds `generated_hltb_main_story` the same unguarded way, so it gets the same column guard.
- `full_path_digest_sql` now coalesces a NULL `fs_path`/`fs_name`, matching `compute_full_path_hash`. Both columns are declared NOT NULL, but where the two spellings disagreed the SQL folded the whole row to NULL and the backfill would abort `0126` at the `NOT NULL` step, which is exactly how a database ends up in the state above.

Both migrations were already applied in an alpha, so their revision ids are unchanged: an install that got through them sees no-ops, and an install stuck mid-`0126` completes it.

AI assistance disclosure, per `CONTRIBUTING.md`: this change was written with Claude Code (diagnosis, patch, tests and this description), reviewed before submission.

**Checklist**
<sup>Please check all that apply.</sup>

- [x] I've tested the changes locally
- [x] I've updated relevant comments
- [ ] I've assigned reviewers for this PR
- [x] I've added unit tests that cover the changes

Three tests in `backend/tests/test_migrations.py`, all passing under `pytest (mariadb)` and `pytest (postgresql)`:

- `test_a_migration_replayed_over_a_migrated_schema_is_a_no_op`, parametrized over `0126` and `0128` — the replay that produces the duplicate-column error today.
- `test_the_full_path_hash_migration_resumes_an_interrupted_run` builds the real intermediate state (downgrade, then re-add the bare column): column present with no digest, no `NOT NULL`, neither index rebuilt. Asserts the replay completes all three and backfills the digest the model computes in Python.
- `test_the_hltb_migration_resumes_an_interrupted_run` drops `0128`'s index, keeps the generated column, and asserts the replay restores it.

Plus a NULL case on the existing digest parity test.

The migration chain itself is exercised end to end by `migrate-mariadb` (10.11.19 and 12.3.3, `utf8mb4_general_ci` and `utf8mb3_unicode_ci`) and `migrate-postgres`, all green.

#### Screenshots (if applicable)

n/a

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WRP83vJP7Cws6SSdE3XZ1a

### CODE REVIEW

File: backend/tests/test_migrations.py [hunk 1]
```diff
@@ -101,3 +108,33 @@ def test_the_migrated_full_path_digest_matches_the_models(fs_path: str, fs_name:
...
+    with sync_engine.begin() as connection:
+        with Operations.context(MigrationContext.configure(connection)):
+            migration.upgrade()
+
+        indexes = {
+            index["name"]: index for index in sa.inspect(connection).get_indexes("roms")
+        }
+
+    assert indexes[migration.UNIQUE_INDEX_NAME]["unique"]
+    assert not indexes[migration.LOOKUP_INDEX_NAME]["unique"]
```
[the remark this answers is not in the document]
Comment (sdornan) [reply backend/tests/test_migrations.py]: Fair, and fixed in 955958a2.

Partly fair, to be precise: the replay over a completed schema does exercise the column guard, because that is the statement the unguarded revision dies on ("Duplicate column name 'full_path_hash'"). What it did not exercise is the rest of a half-applied run, and 0128 had no replay coverage at all.

There are now three tests:

- `test_a_migration_replayed_over_a_migrated_schema_is_a_no_op`, parametrized over 0126 and 0128.
- `test_the_full_path_hash_migration_resumes_an_interrupted_run` builds the real intermediate state (downgrade, then re-add the bare column) so the column is present with no digest, no NOT NULL and neither index rebuilt, then asserts the replay completes all three and backfills the digest the model computes in Python.
- `test_the_hltb_migration_resumes_an_interrupted_run` drops 0128's index, keeps the generated column, and asserts the replay restores it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: backend/alembic/versions/0128_hltb_main_story_column.py [hunk 2]
```diff
@@ -47,11 +48,18 @@
...
-    op.execute(  # nosec B608
-        f"ALTER TABLE roms ADD COLUMN {COLUMN_NAME} BIGINT "
-        f"GENERATED ALWAYS AS ({expr}) STORED"
-    )
+    connection = op.get_bind()
+    columns = {column["name"] for column in sa.inspect(connection).get_columns("roms")}
+
+    # MySQL/MariaDB auto-commit each DDL statement, so a run that dies on the
+    # index keeps the column without advancing the alembic version.
+    if COLUMN_NAME not in columns:
```
[the remark this answers is not in the document]
Comment (sdornan) [reply backend/alembic/versions/0128_hltb_main_story_column.py:56]: Added in 955958a2: `test_the_hltb_migration_resumes_an_interrupted_run` drops `idx_roms_hltb_main_story`, leaves the generated column in place, and replays 0128, asserting the index comes back. An unguarded `ADD COLUMN` fails that test on the duplicate column. 0128 is also covered by the parametrized replay-over-a-complete-schema test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: backend/tests/test_migrations.py [hunk 3]
```diff
@@ -101,3 +108,33 @@ def test_the_migrated_full_path_digest_matches_the_models(fs_path: str, fs_name:
         ).scalar_one()
 
     assert digest == compute_full_path_hash(fs_path, fs_name)
+
+
+def test_the_full_path_hash_migration_survives_a_re_run():
+    """0126 replayed over a schema it already migrated is a no-op.
+
+    MySQL/MariaDB auto-commit each DDL statement, so a run that dies partway
+    keeps the column without advancing the alembic version, and every restart
```
[the remark this answers is not in the document]
Comment (sdornan) [reply backend/tests/test_migrations.py]: Both done in 955958a2: the replay test is now parametrized over 0126 and 0128, and a separate test builds the partial schema (bare nullable column, no digest, neither index) and asserts the replay finishes it, including the backfill.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by gantoine** [comment]
Superseded by #4484, which carries these fixes in full alongside the same class of bug in 0111, 0119, 0120, 0122 and 0123.
